src/darwin/Framework/CHIP/MTROTAHeaderParser.mm

```diff
@@ -0,0 +1,72 @@
+/**
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTROTAHeaderParser.h"
+
+#import "MTRError.h"
+#import "MTRError_Internal.h"
+
+#include <lib/core/OTAImageHeader.h>
+
+@implementation MTROTAHeaderParser
++ (nullable NSDictionary *)dictFromData:(NSData *)data error:(NSError * __autoreleasing *)error
+{
+    chip::OTAImageHeaderParser parser;
+
+    parser.Init();
+
+    if (!parser.IsInitialized()) {
+        *error = [NSError errorWithDomain:MTRErrorDomain code:MTRErrorCodeGeneralError userInfo:nil];
+        return nil;
+    }
+
+    chip::ByteSpan buffer(static_cast<const unsigned char *>([data bytes]), [data length]);
+    chip::OTAImageHeader header;
+
+    CHIP_ERROR err = parser.AccumulateAndDecode(buffer, header);
+    if (err != CHIP_NO_ERROR) {
+        *error = [MTRError errorForCHIPErrorCode:err];
+        parser.Clear();
+        return nil;
+    }
+
+    auto headerDict = [[NSMutableDictionary alloc] init];
+    headerDict[@"vendorId"] = @(header.mVendorId);
+    headerDict[@"productId"] = @(header.mProductId);
+    headerDict[@"payloadSize"] = @(header.mPayloadSize);
+    headerDict[@"softwareVersion"] = @(header.mSoftwareVersion);
+    headerDict[@"softwareVersionString"] = [[NSString alloc] initWithBytes:header.mSoftwareVersionString.data()
```

Comment:
I wonder whether we should have a helper for converting CharSpan to NSString....

Reply:
I think this would be a good idea, because it may contain junk at the end or I would suggest to do the following:
`[[NSString alloc] initWithBytes:header.mSoftwareVersionString.data() length:header.mSoftwareVersionString.size() encoding:NSUTF8StringEncoding]`

Reply:
I have added one.